node/impl/full/eth.go

```diff
@@ -241,7 +241,68 @@ func (a *EthModule) EthGetBlockByHash(ctx context.Context, blkHash ethtypes.EthH
 	return newEthBlockFromFilecoinTipSet(ctx, ts, fullTxInfo, a.Chain, a.StateAPI)
 }
 
-func (a *EthModule) parseBlkParam(ctx context.Context, blkParam string, strict bool) (tipset *types.TipSet, err error) {
+func (a *EthModule) getTipsetByEthBlockNumberOrHash(ctx context.Context, blkParam ethtypes.EthBlockNumberOrHash) (*types.TipSet, error) {
+	head := a.Chain.GetHeaviestTipSet()
+
+	predefined := blkParam.PredefinedBlock
+	if predefined != nil {
+		if *predefined == "earliest" {
+			return nil, fmt.Errorf("block param \"earliest\" is not supported")
+		} else if *predefined == "pending" {
+			return head, nil
+		} else if *predefined == "latest" {
+			parent, err := a.Chain.GetTipSetFromKey(ctx, head.Parents())
+			if err != nil {
+				return nil, fmt.Errorf("cannot get parent tipset")
+			}
+			return parent, nil
+		} else {
+			return nil, fmt.Errorf("unknown predefined block %s", *predefined)
+		}
+	}
+
+	if blkParam.BlockNumber != nil {
+		height := abi.ChainEpoch(*blkParam.BlockNumber)
+		if height > head.Height()-1 {
+			return nil, fmt.Errorf("requested a future epoch (beyond 'latest')")
+		}
+		ts, err := a.ChainAPI.ChainGetTipSetByHeight(ctx, height, head.Key())
+		if err != nil {
+			return nil, fmt.Errorf("cannot get tipset at height: %v", height)
+		}
+		return ts, nil
+	}
+
+	if blkParam.BlockHash != nil {
+		ts, err := a.Chain.GetTipSetByCid(ctx, blkParam.BlockHash.ToCid())
+		if err != nil {
+			return nil, fmt.Errorf("cannot get tipset by hash: %v", err)
+		}
+
+		if blkParam.RequireCanonical {
+			// walk back the current chain (our head) until we reach targetHeight and validate the block hash
+			currTs := head
+			for {
+				if currTs.Equals(ts) {
+					return ts, nil
+				} else if currTs.Height() < ts.Height() {
+					return nil, fmt.Errorf("could not find block hash %s in canonical chain", blkParam.BlockHash.ToCid())
+				}
+
+				currTs, err = a.Chain.LoadTipSet(ctx, currTs.Parents())
+				if err != nil {
+					return nil, fmt.Errorf("failed to load tipset: %v", err)
+				}
+			}
```

Comment:
Call `ChainGetTipSetByHeight` and let that walk for you. That'll use the index/skip-list.

Reply:
Then, when you have the tipset at the target height:

1. Make sure it's actually at the target height (nulls, etc.). Just in case.
2. Make sure it equals the expected tipset.